src/App.tsx

```diff
@@ -1,26 +1,395 @@
+    });
+
+    if (tempTodo) {
+      setTodos(prevTodos => [...prevTodos, tempTodo]);
+    }
+
+    return todoService
+      .addTodo({
+        userId: todoService.USER_ID,
+        title: input.trim(),
+        completed: false,
+      })
+      .then(newTodo => {
+        setTodos(prevTodos =>
+          prevTodos.filter(todo => todo.id !== tempTodo?.id),
+        );
+
+        setTodos(prevTodos => [...prevTodos, newTodo]);
+        setInput('');
+        setTempTodo(null);
+      })
+      .catch(error => {
+        setTodos(todos);
+        setTempTodo(null);
+        setErrorMessage(Error.UnableToAdd);
+        throw error;
+      })
+      .finally(() => setIsSubmitting(false));
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setErrorMessage('');
+    setIsSubmitting(true);
+    setLoadingTodoIds(prev => [...prev, todoId]);
+
+    return todoService
+      .deleteTodo(todoId)
+      .then(() =>
+        setTodos(prevTodos =>
+          prevTodos.filter((todo: Todo) => todo.id !== todoId),
+        ),
+      )
+      .catch(error => {
+        setTodos(todos);
+        setErrorMessage(Error.UnableToDelete);
+        throw error;
+      })
+      .finally(() => {
+        setLoadingTodoIds(prev => prev.filter(id => id !== todoId));
+        setIsSubmitting(false);
+      });
+  };
+
+  const handleClearCompleted = () => {
+    setErrorMessage('');
+    setIsSubmitting(true);
+
+    const deletePromises = completedTodos.map(todo => {
+      setLoadingTodoIds(prev => [...prev, todo.id]);
+
+      return todoService.deleteTodo(todo.id);
+    });
+
+    Promise.allSettled(deletePromises)
+      .then(results => {
+        const deletingIds: number[] = [];
+
+        results.forEach((result, index) => {
+          if (result.status === 'fulfilled') {
+            deletingIds.push(completedTodos[index].id);
+          } else {
+            setErrorMessage(Error.UnableToDelete);
+          }
+        });
+
+        deletingIds.forEach(deletingId => {
+          setTodos(prevTodos =>
+            prevTodos.filter(todo => todo.id !== deletingId),
+          );
+        });
+      })
+      .catch(error => {
+        setTodos(todos);
+        setErrorMessage(Error.UnableToDelete);
+        throw error;
+      })
+      .finally(() => {
+        setLoadingTodoIds([]);
+        setIsSubmitting(false);
+      });
+  };
+
+  const handleEditTodo = (currentTodo: Todo) => {
+    setErrorMessage('');
+    setIsSubmitting(true);
+    setLoadingTodoIds(prev => [...prev, currentTodo.id]);
+
+    return todoService
+      .updateTodo(currentTodo.id, currentTodo.title.trim())
+      .then(updatedTodo => {
+        setTodos(prevTodos =>
+          prevTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          ),
+        );
+        setServerError(false);
+      })
+      .catch(error => {
+        setTodos(todos);
+        setErrorMessage(Error.UnableToUpdate);
+        setServerError(true);
+        throw error;
```

Comment:
In case you throw error here no need to `setServerError ` and pass it as a prop, just catch it in the `TodoItem` (or wherever you call this function) and handle additionally

Reply:
same for all similar cases

Reply:
serverError is used in TodoItem not to close Editing Form if there is an error